plugins/formal-verification/formver.core/src/org/jetbrains/kotlin/formver/domains/Injection.kt

```diff
@@ -0,0 +1,43 @@
+/*
+ * Copyright 2010-2024 JetBrains s.r.o. and Kotlin Programming Language contributors.
+ * Use of this source code is governed by the Apache 2.0 license that can be found in the license/LICENSE.txt file.
+ */
+
+package org.jetbrains.kotlin.formver.domains
+
+import org.jetbrains.kotlin.formver.domains.UnifiedTypeDomain.Companion.isOf
+import org.jetbrains.kotlin.formver.viper.ast.*
+
+/**
+ * Produces set of axioms for injections from built-in types in Viper to Ref.
+ *
+ * @param viperType: built-in type which needs to be mapped
+ * @param typeFunction: representation of that type as a domain func
+ */
+class Injection(
+    injectionName: String,
+    val viperType: Type,
+    val typeFunction: DomainFunc
+) {
+    private val v = Var("v", viperType)
+    private val r = Var("r", Type.Ref)
+    val toRef = UnifiedTypeDomain.createDomainFunc("${injectionName}ToRef", listOf(v.decl()), Type.Ref)
+    val fromRef = UnifiedTypeDomain.createDomainFunc("${injectionName}FromRef", listOf(r.decl()), viperType)
+
+    internal fun AxiomListBuilder.injectionAxioms() {
+        axiom {
+            Exp.forall(v) { v -> simpleTrigger { toRef(v) isOf typeFunction() } }
+        }
+        axiom {
+            Exp.forall(v) { v ->
+                simpleTrigger { fromRef(toRef(v)) } eq v
+            }
+        }
+        axiom {
+            Exp.forall(r) { r ->
+                assumption { r isOf typeFunction() }
+                simpleTrigger { toRef(fromRef(r)) } eq r
+            }
+        }
```

Comment:
Perhaps this should be conditional on `viperType` being `Ref`, in which case we can have `toRef(v) = v` and `fromRef(v) = v`?

Reply:
I think we never should have an injection for `Ref`, especially with this design which accepts type function...

Reply:
But then will this axiom ever be triggered?